src/org/javarosa/xpath/expr/XPathFuncExpr.java

```diff
@@ -344,6 +344,16 @@ public Object eval (DataInstance model, EvaluationContext evalContext) {
             }
         } else if (name.equals("substr") && (args.length == 2 || args.length == 3)) {
             return substring(argVals[0], argVals[1], args.length == 3 ? argVals[2] : null);
+        } else if (name.equals("substring-before") && args.length == 2) {
+            String str = toString(argVals[0]);
```

Comment:
I'd like to suggest some changes:
- [MDN](https://developer.mozilla.org/en-US/docs/Web/XPath/Functions/substring-before) suggests `haystack` and `needle` argument names. I think those names do a great work of explaining what's going on.
- Use `needleIndex` instead of `pos`. It's not abbreviated, it's more expressive (you're calling an `indexOf` method, therefore, the result should be an index), and it's explicit about its relationship with the `needle`.
- Remove the parenthesis (redundant) in the `return` line.
- I think the comment should be moved to a line before the return and explicitly say that the XPath reference states that an empty string should be returned when the `needle` isn't found. Otherwise, it could look like some arbitrary decision on our side.

```java
String haystack = toString(argVals[0]);
String needle = toString(argVals[1]);
int needleIndex = haystack.indexOf(needle);
// XPath reference states that we should return the empty string when we don't find the needle in the haystack
return needleIndex >= 0 ? haystack.substring(0, needleIndex) : "";
```

Reply:
Thanks for your suggestions. Done.

Reply:
I think the needle in a haystack metaphor is dumb, and phooey on MDN for using it.

https://dictionary.cambridge.org/us/dictionary/english/a-needle-in-a-haystack

Reply:
I have no opinion either way... But I do notice that javaRosa does sometime errs on the side of 'verbosity', so its not completely out of character. eg

return GeoUtils.calculateAreaOfGPSPolygonOnEarthInSquareMeters(latLongs);

;-)

Reply:
Ha ha.
Verbosity is a different dimension from frivolity or suitability, thought. 
Was it a coincidence that you picked something *I* named? ;-)

Reply:
>Was it a coincidence that you picked something I named?

D'oh! :-)

Reply:
> I think the needle in a haystack metaphor is dumb, and phooey on MDN for using it.

Huh, I didn't know about that. This seemed to be an improvement over `arg0` and `arg1`, but now I'm hesitating. Do you have any better alternatives, @dcbriccetti?